File: GaniOSApp/GanClient/GanClient/Gan/Service/SystemService.swift
//
//  SystemService.swift
//  GanClient
//
//  Created by xattacker on 2020/11/23.
//

import Foundation


public final class SystemService: ServiceFoundation
{
    public func getIP() -> String?
    {
        var ip: String?
        
        switch self.send(FunctionType.get_ip)
        {
            case .success(let response):
                if response.result
                {
                    ip = response.responseString
                }
                break
                
            case .failure(let error):
                print(error)
                break
        }

        return ip
    }
    
    public func getSystemTime() -> Date?
    {
        var time: Date?
        
        switch self.send(FunctionType.get_system_time)
        {
            case .success(let response):
                if response.result,
                   let data = response.response,
                   let timestamp = BinaryBuffer(data: data).readLongLong()
                {
                    time = Date(timestamp: timestamp)
                }
                break
                
            case .failure(let error):
                print(error)
                break
        }

        return time
    }
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/gan/service/SystemService.kt
package com.xattacker.gan.service

import com.xattacker.binary.TypeConverter
import com.xattacker.gan.GanAgent
import com.xattacker.gan.data.FunctionType
import java.util.*


class SystemService internal constructor(agent: GanAgent) : ServiceFoundation(agent)
{
    fun getIP(): String?
    {
        var ip: String? = null
        try
        {
            val response = send(FunctionType.GET_IP, null)
            if (response != null && response.result && response.response != null)
            {
                    ip = String(response.response!!)
            }
        }
        catch (ex: Exception)
        {
            ip = null
        }

        return ip
    }

    fun getSystemTime(): Date?
    {
        var time: Date? = null
        try
        {
            val response = send(FunctionType.GET_SYSTEM_TIME, null)
            if (response != null && response.result && response.response != null)
            {
                val timestamp: Long = TypeConverter.byteToLong(response.response!!)
                time = Date(timestamp)
            }
        }
        catch (ex: Exception)
        {
            time = null
        }

        return time
    }
}